[quant][fx][graphmode][api] Change API for custom module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45920 [quant][fx][graphmode][api] Change API for custom module**

Summary:
See docs for new way of defining custom modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24145856](https://our.internmc.facebook.com/intern/diff/D24145856)

jerryzh168 committed:
[quant][fx][graphmode][api] Change API for custom module

Summary:
See docs for new way of defining custom modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode][api] Change API for custom module"


Summary:
See docs for new way of defining custom modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24145856](https://our.internmc.facebook.com/intern/diff/D24145856)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode][api] Change API for custom module"


Summary:
See docs for new way of defining custom modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24145856](https://our.internmc.facebook.com/intern/diff/D24145856)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -717,26 +714,14 @@ def forward(self, x):
-            qconfig_dict = {'': default_qconfig}
+            qconfig_dict = {
+                '': default_qconfig,
+                'custom_module_class':
```
Do we want custom and standalone modules to be part of the qconfig dict?  Isnt this making the qconfig dict more complex? Can we handle CustomModule by defining the module mappings separately? Also, can standalone modules be a separate argument? This way qconfigDict only specifies qconfigs, not module mappings or other info.

jerryzh168 replied:
we can discuss, but I think it's easier for users to specify everything in qconfig_dict so that they don't need to remember other apis. these are optional arguments as well, only people who need this feature need to know about it.

vkuzo replied:
seems like currently (before this PR):
* key of qconfig dict defines the subgraph the rule applies to
* value of qconfig dict defines the observers used for the subgraph matching the key

I think it could be reasonable to make qconfig hold more information other than the observers, my question would be if we can make sure each type of information is only held in one place.  For example, if we are adding custom mappings to qconfig_dict, it would be ideal to make qconfig_dict the only place where custom mappings could be defined.  @jerryzh168 , are there any other ways to customize things right now other than the qconfig_dict?

jerryzh168 replied:
currently for custom modules user can call `register_observed_custom_module_mapping`, but we can make that api private

jerryzh168 committed:
Update on "[quant][fx][graphmode][api] Change API for custom module"


Summary:
See docs for new way of defining custom modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24145856](https://our.internmc.facebook.com/intern/diff/D24145856)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode][api] Change API for custom module"


Summary:
See docs for new way of defining custom modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24145856](https://our.internmc.facebook.com/intern/diff/D24145856)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode][api] Change API for custom module"


Summary:
See docs for new way of defining custom modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24145856](https://our.internmc.facebook.com/intern/diff/D24145856)

[ghstack-poisoned]

vkuzo reviewed (approved):
chatted offline, we will iterate on future PRs, lg!

jerryzh168 committed:
Update on "[quant][fx][graphmode][api] Change API for custom module"


Summary:
See docs for new way of defining custom modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24145856](https://our.internmc.facebook.com/intern/diff/D24145856)

[ghstack-poisoned]
